### PR TITLE
Factor out common limits handling for x/y/z axes.

### PR DESCRIPTION
**Pull request by anntzer**
The change in set_top_view is necessary because super().set_xlim no
longer goes to 2D semantics (it's on the Axis.set_view_limits that one
would need to do a supercall), but in any case the old behavior was
actually more buggy because it would e.g. confuse data-limits sharex and
2D projection sharex; e.g.
`subplots(1, 2, sharex=True, subplot_kw={"projection": "3d"})`
was previously wrong.

As a side effect, also fixes `invert_yaxis`/`invert_zaxis` on 3d axises (#21369)
(because they just go to the base `Axis.set_inverted`).  Redoing the inheritance
hierarchy (#21424) would probably(?) still be nice, at least from a semantic PoV.

old bad 3d sharex:
old

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -1012,6 +1013,96 @@ def set_default_intervals(self):
         # attribute, and the derived code below will check for that
         # and use it if it's available (else just use 0..1)
 
+    def _set_lim(self, v0name, v1name, v0, v1, alt0=None, alt1=None, *,
+                 emit=True, auto):
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:1017]: The signature is too complex and abstracting away the direction makes it even more hard to understand. Since this is internal, we don't have to jump all the hoops of the public set_*lim API. What about:

```suggestion
    def _set_lim(self, v0, v1, vnames=None, *, emit=True, auto):
```
where `v0`, `v1` can optionally be 2-tuples.

This would change the xlim case from:

```
        return self.xaxis._set_lim(
            "left", "right", left, right, xmin, xmax, emit=emit, auto=auto)
```
to
```
        return self.xaxis._set_lim(
            (xmin, left), (ymin, right), lim_names=("left", "right"), emit=emit, auto=auto)
```

Comment (anntzer) [reply lib/matplotlib/axis.py:1017]: I could group vnames together, but grouping v0 and alt0 would make the logic in set_lim even worse, because now v0 being a tuple could come either from grouping v0 and alt0 or from v0 being passed in as a (v0, v1) pair...  (unless you want to have the user *always* pass in both v0 and alt0, but that would complicate the implementation of set_inverted.)

Comment (timhoffm) [reply lib/matplotlib/axis.py:1017]: I see.

What about
```suggestion
    def _set_lim(self, v0, v1, alt0=None, alt1=None, *, vnames=None,
                 emit=True, auto):
```

Comment (anntzer) [reply lib/matplotlib/axis.py:1017]: I'm not really sure the first change helps either; e.g. I'll need to immediately unpack `vnames` inside the implementation anyways, and then you're just making the call sites longer if you force an explicit kwarg, even though it's quite clear that `set_lim("left", "right", ...)` is passing argnames.

I can't say I particularly like using e.g. xaxis names here (I don't have a very strong argument against, other than "v0 seems just fine to me"), but I'll let others chime in and can change that if they prefer them too.

Comment (timhoffm) [reply lib/matplotlib/axis.py:1017]: > it's quite clear that `set_lim("left", "right", ...)` is passing argnames.

I found that very confusing the first time I saw it. I’m coming from set_slim and was expecting two numbers as the first two arguments. The two strings are quite obscure.

I have no strong opinion on x vs. v. Just wanted to mention the possibility.

Comment (anntzer) [reply lib/matplotlib/axis.py:1017]: Ah, fair enough.  Renamed to `names`.

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -1012,6 +1012,97 @@ def set_default_intervals(self):
...
+        *names* is the pair of the names of the first two parameters of the
+        Axes method (e.g., "left" and "right").  They are only used to generate
+        error messages; and can be empty if the limits are known to be valid.
+
+        Other parameters are directly forwarded from the Axes limits setter:
+        *v0*, *v1*, *alt0*, and *alt1* map to *left*, *right*, *xmin*, and
+        *xmax* respectively; *emit* and *auto* are used as is.
+        """
+        v0name, v1name = names  # The value names.
+        name, = [name for name, axis in self.axes._get_axis_map().items()
```
Comment (jklymak) [inline lib/matplotlib/axis.py:1033]: This seems a really obscure way to get the name to me.  Is there some reason an axis can't know its "name" when it is created?

**Review by jklymak — COMMENTED** [review]
Overall :+1:  - centralizing this makes sense, even if the root code is a little hared to parse, but the comments all help.  Needs a rebase, and maybe a bit of thought about where "names" comes from - feel free to ping for a re-review.

File: lib/matplotlib/axis.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/axis.py:1044]: Axises know their "name", but that's e.g. "r"/"theta" for polar axes whereas here we really want "x"/"y" even for polar.  Perhaps Axises should have multiple name attributes for the various use cases, but that would be a bit hairy too...

**Comment by anntzer** [comment]
Rebased.

**Review by jklymak — APPROVED** [review]

File: lib/matplotlib/axis.py [hunk 3]
```diff
@@ -1023,6 +1023,97 @@ def set_default_intervals(self):
...
+        if alt0 is not None:
+            if v0 is not None:
+                raise TypeError(
+                    f"Cannot pass both {v0name!r} and '{name}lim'")
+            v0 = alt0
+        if alt1 is not None:
+            if v1 is not None:
+                raise TypeError(
+                    f"Cannot pass both {v1name!r} and '{name}lim'")
+            v1 = alt1
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:1058]: I'm wondering whether this must be part of an internal `_set_lim()` function. This is only needed to support some weird parameter duplication in the calling functions. Can we leave that part in the calling functions? That would keep it more readable the variable mapping is hard to follow. Alternatively, could we have a decorator

```
@_api.alias(left='xlim', right='ylim')
def set_xlim(self, left=None, right=None, emit=True, auto=False):
```
that does the checking? Then, the aliases don't show in the signature, but I would consider that a plus. IMHO it's better to have the signature only illustrate one case, and use the docstring to say "You can use 'xmin' instead of 'left' and ...". That's a bit peculiar but then again, the duplication is quite confusing as well.

Comment (anntzer) [reply lib/matplotlib/axis.py:1058]: I can move back the parameter dealiasing to set_{x,y,z}lim and duplicate them 3x again, but I'm not sure if this really helps...  (In particular, you still have to pass `names`, which are used further down as well)  Also, moving that logic to a separate `_api.alias` also seems more complexity than needed (in particular by splitting the logic to two different functions/files).

I do agree that the signature can be clearer, but that can be done with `set_xlim.__signature__ = inspect.signature(lambda self, left=None, right=None, emit=True, auto=False: None)` (which is probably a good idea, but can be done separately).

Comment (timhoffm) [reply lib/matplotlib/axis.py:1058]: > I can move back the parameter dealiasing to set_{x,y,z}lim and duplicate them 3x again, but I'm not sure if this really helps...

I think yes. It's some structural reduncancy, but the names are always different and having the original names and the context of the docstring helps.

Additionally, it's a great simplification if `set_lim` only has to bother with two values for the limits instead of four. Even though two of them are only used at the top, the function is so long that you can see that directly.

You may even get away without the names by re-wording a little. See below.

File: lib/matplotlib/axis.py [hunk 4]
```diff
@@ -1023,6 +1023,97 @@ def set_default_intervals(self):
...
+                v1 = old1
+
+        if self.get_scale() == 'log' and (v0 <= 0 or v1 <= 0):
+            # Axes init calls set_xlim(0, 1) before get_xlim() can be called,
+            # so only grab the limits if we really need them.
+            old0, old1 = self.get_view_interval()
+            if v0 <= 0:
+                _api.warn_external(
+                    f"Attempt to set non-positive {v0name} {name}lim on a "
+                    f"log-scaled axis will be ignored.")
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:1080]: Is {v0name} really needed here?

> Attempt to set non-positive xlim on a log-scaled axis

is almost as good as 

> Attempt to set non-positive left xlim on a log-scaled axis

I suppose the user can still identify where the problem is.

File: lib/matplotlib/axis.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/axis.py:1058]: OK, done (as a separate commit).

The "canonical" property names for Axes3D were switched from xlim3d/ylim3d to xlim/ylim so that Axes3D can directly inherit the setters from 2D Axes (avoiding more duplication there).

File: lib/matplotlib/axis.py [hunk 5]
```diff
@@ -1023,6 +1023,76 @@ def set_default_intervals(self):
         # attribute, and the derived code below will check for that
         # and use it if it's available (else just use 0..1)
 
+    def _set_lim(self, v0, v1, *, emit=True, auto):
+        """
+        Set view limits.
+
+        This method is a helper for the Axes ``set_xlim``, ``set_ylim``, and
+        ``set_zlim`` methods.  This docstring uses names corresponding to
+        ``set_xlim`` for simplicity.
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:1032]:
```suggestion
        ``set_zlim`` methods.
```
This is no longer the case.

File: lib/matplotlib/axis.py [hunk 6]
```diff
@@ -1023,6 +1023,76 @@ def set_default_intervals(self):
...
+
+        *v0* and *v1* are the view limits; they must be two scalars (passing a
+        *v0* as a (low, high) pair is not supported).  *emit* and *auto* are
+        forwarded as-is from the Axes limits setter.
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:1036]: I think it's better to have a standard parameters section,  and not write the docstring with the perspective of the calling functions in mind.

```suggestion
        Parameters
        ----------
        v0, v1 : float
            The view limits.
        emit : bool, default: True
            Whether to notify observers of limit change.
        auto : bool or None
            Whether to turn on autoscaling of the x-axis. True turns on,
            False turns off, None leaves unchanged.
```

File: lib/matplotlib/axis.py (hunk 5 above)
Comment (timhoffm) [reply lib/matplotlib/axis.py:1032]: Oh, the docstring (and signature) don't have that notion, but some comments in the function still have.

File: lib/matplotlib/axis.py [hunk 7]
```diff
@@ -1023,6 +1023,76 @@ def set_default_intervals(self):
...
+                                   f"a log-scaled axis will be ignored.")
+                v1 = old1
+        if v0 == v1:
+            _api.warn_external(
+                f"Attempting to set identical low and high {name}lims "
+                f"makes transformation singular; automatically expanding.")
+        reverse = v0 > v1
+        v0, v1 = self.get_major_locator().nonsingular(v0, v1)
+        v0, v1 = self.limit_range_for_scale(v0, v1)
+        # cast to bool to avoid bad interaction between python 3.8 and np.bool_
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:1073]: Optional drive-by improvement: It would be more logical to move the comment and casting to the definition of `reverse` above.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 8]
```diff
@@ -785,62 +671,29 @@ def set_zlim3d(self, bottom=None, top=None, emit=True, auto=False,
...
+    set_ylim3d = maxes.Axes.set_ylim
+    set_zlim3d = set_zlim
 
-    def get_xlim3d(self):
+    def get_xlim(self):
         return tuple(self.xy_viewLim.intervalx)
-    get_xlim3d.__doc__ = maxes.Axes.get_xlim.__doc__
 
-    def get_ylim3d(self):
+    get_xlim.__doc__ = maxes.Axes.get_xlim.__doc__
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:689]: Why do you need this? Isn't the docstring inherited from `Axes` anyway?

File: lib/matplotlib/axis.py (hunk 6 above)
Comment (anntzer) [reply lib/matplotlib/axis.py:1036]: sure

File: lib/matplotlib/axis.py (hunk 7 above)
Comment (anntzer) [reply lib/matplotlib/axis.py:1073]: done

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 8 above)
Comment (anntzer) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: good catch, this became unneeded now that xlim3d is an alias of xlim rather than the other way round.

File: lib/matplotlib/axis.py [hunk 9]
```diff
@@ -1004,10 +1004,10 @@ def set_inverted(self, inverted):
         the top for the y-axis; the "inverse" direction is increasing to the
         left for the x-axis and to the bottom for the y-axis.
         """
-        # Currently, must be implemented in subclasses using set_xlim/set_ylim
-        # rather than generically using set_view_interval, so that shared
-        # axes get updated as well.
-        raise NotImplementedError('Derived must override')
+        # docstring inherited
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:1007]: You can't have that comment and a docstring. :)

**Review by timhoffm — APPROVED** [review]
May self-merge after adressing the one remaining comment.

File: lib/matplotlib/axis.py (hunk 9 above)
Comment (anntzer) [reply lib/matplotlib/axis.py]: oops, fixed.

**Comment by timhoffm** [comment]
Doc failures are real.

**Comment by anntzer** [comment]
Test failure should be fixed now; foc failures seem related to the new scipy 1.8 release?

**Comment by QuLogic** [comment]
This broke docs (see #22431):
```
.../lib/mpl_toolkits/mplot3d/axes3d.py:docstring of mpl_toolkits.mplot3d.axes3d.Axes3D.get_xlim:20: WARNING: py:obj reference target not found: set_xlim
.../lib/mpl_toolkits/mplot3d/axes3d.py:docstring of mpl_toolkits.mplot3d.axes3d.Axes3D.get_ylim:20: WARNING: py:obj reference target not found: set_ylim
```
